docs/azure/sdk/protocol-convenience-methods.md

```diff
@@ -0,0 +1,132 @@
+
+> [!NOTE]
+> The `Azure.Core` library itself also depends on `System.ClientModel` for various service building blocks. In the context of this article, the key differentiator for method patterns is whether a client library depends on `Azure.Core` or `System.ClientModel` directly, rather than through a transitive dependency.
+
+The following table compares some of the request and response types used by protocol and convenience methods, based on whether the library depends on `Azure.Core` or `System.ClientModel`:
+
+|Request or response concern  |Azure.Core  | System.ClientModel |
+|---------|---------|---------|
+|Request body     | [`RequestContent`](/dotnet/api/azure.core.requestcontent)         | [`BinaryContent`](/dotnet/api/system.clientmodel.binarycontent)         |
+|Advanced options     | [`RequestContext`](/dotnet/api/azure.requestcontext)         | [`RequestOptions`](/dotnet/api/system.clientmodel.primitives.requestoptions)         |
+|Raw HTTP Response     | [`Response`](/dotnet/api/azure.response)         | [`PipelineResponse`](/dotnet/api/system.clientmodel.primitives.pipelineresponse)         |
+|Return type with output model     | [`Response<T>`](/dotnet/api/azure.response-1)         | [`ClientResult<T>`](/dotnet/api/system.clientmodel.clientresult-1)         |
+
+The sections ahead provide implementation examples of these concepts.
+
+## Protocol and convenience method examples
+
+The coding patterns and types used by client library protocol and convenience methods vary slightly based on whether the library depends on `Azure.Core` or `System.ClientModel`. The differences primarily influence the .NET types used for handling request and response data.
+
+### Libraries that depend on Azure.Core
+
+Many Azure SDK client libraries depend on the `Azure.Core` library. For example, the [`Azure.AI.ContentSafety`](/dotnet/api/overview/azure/ai.contentsafety-readme) library depends on the `Azure.Core` library and provides a `ContentSafetyClient` class that exposes both protocol and convenience methods.
+
+### [Protocol method](#tab/protocol-methods)
+
+The following code uses a `ContentSafetyClient` to call the `AnalyzeText` protocol method:
+
+:::code source="snippets/protocol-convenience-methods/AzureCoreProtocol/Program.cs" highlight="18-20":::
+
+The preceding code demonstrates the following protocol method patterns:
+
+- Uses the `RequestContent` type to supply data for the request body.
+- Uses the `RequestContext` type to configure request options.
+- Returns data using the `Response` type.
+- Reads the `ContentStream` to access the response data.
+
+> [!NOTE]
+> The preceding code configures the `ClientErrorBehaviors.NoThrow` for the `RequestOptions`. This option prevents non-success service responses status codes from throwing an exception, which means the app code should manually handle the response status code checks.
```

Comment:
Would it make sense to show this in the sample?  e.g.

```diff
// Call the protocol method
Response response = safetyClient.AnalyzeText(
    prompt, 
    new RequestContext() { ErrorOptions = ErrorOptions.NoThrow});

+ if (response.Status != 200)
+ {
+  throw new RequestFailedException(response);
+ }

using (StreamReader streamReader = new StreamReader(response.ContentStream))
{
    // Display the results
    Console.WriteLine(streamReader.ReadToEnd());
}
```

Reply:
Wouldn't this imply that other 200s such as 202 would be an exception? I think that might be a bit strange as a sample, though I agree the current state is also a little odd. I added a note about handling status code, but we could also be more up front with a comment such as "This option is only set for example purposes" or something.

Reply:
Yes, good observation.  The way our generators work is that in service API specifications, successful response codes are listed for each endpoint.  A service might say that it will return `200` or `202` as a success response, but say, `201` is missing from the list of success responses.  If for some reason a response with a `201` status code was returned, the client's convenience method would throw a `RequestFailedException` (or `ClientResultException` in SCM-based clients) because this is not a stated success code.  It is intended that protocol method users would read the service REST API docs to know which status codes correspond to those success codes listed by the service -- not just HTTP "success" status codes, e.g. 2xx codes.  This is a good example of what we mean when we say RequestContext and RequestOptions are low-level/advanced options types -- the user assumes a lot of responsibility when using them.

For the purposes of the sample, some suggestions include -- adding a comment to the above something like

```diff
+ // Any response code other than 200 from Azure ContentSafety AnalyzeText REST API is not considered a success response
+ // See REST API details at https://azure-ai-content-safety-api-docs.developer.azure-api.net/api-details#api=content-safety-service-2023-10-01&operation=TextOperations_AnalyzeText
+ if (response.Status != 200) 
+ {
+ throw new RequestFailedException(response); 
+ }
```

-- or maybe illustrating a conceptually simpler use of RequestContext/RequestOptions.  I'd suggest [AddHeader](https://learn.microsoft.com/en-us/dotnet/api/system.clientmodel.primitives.requestoptions.addheader?view=azure-dotnet-preview#system-clientmodel-primitives-requestoptions-addheader(system-string-system-string)), but unfortunately it's only on RequestOptions today.  Alternatively, there is [AddPolicy](https://learn.microsoft.com/en-us/dotnet/api/azure.requestcontext.addpolicy?view=azure-dotnet#azure-requestcontext-addpolicy(azure-core-pipeline-httppipelinepolicy-azure-core-httppipelineposition)) or a [CancellationToken](https://learn.microsoft.com/en-us/dotnet/api/azure.requestcontext.cancellationtoken?view=azure-dotnet#azure-requestcontext-cancellationtoken) property on both.